### PR TITLE
mds: split up mdstypes

### PR DESCRIPTION
**Pull request by dparmar18**
mdstypes.h contains both MDS-exclusive and client-shared structs.
This PR splits it up into "mdstypes.h" and "sharedtypes.h".

Fixes: https://tracker.ceph.com/issues/3998

Signed-off-by: dparmar18 <dparmar@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/include/cephfs/sharedtypes.h [hunk 1]
```diff
@@ -0,0 +1,946 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
```
Comment (batrick) [inline src/include/cephfs/sharedtypes.h:1]: Missing LICENSE/copyright notice.

File: src/include/cephfs/sharedtypes.h [hunk 2]
```diff
@@ -0,0 +1,946 @@
...
+// vim: ts=8 sw=2 smarttab
+#ifndef CEPH_SHAREDTYPES_H
```
Comment (batrick) [inline src/include/cephfs/sharedtypes.h:3]: I think this file can be renamed `types.h`. It's already under `cephfs/`.

File: src/mds/mdstypes.h [hunk 3]
```diff
@@ -1137,7 +247,7 @@ inline void decode(old_inode_t<Allocator> &c, ::ceph::buffer::list::const_iterat
 }
 
 /*
- * like an inode, but for a dir frag 
+ * like an inode, but for a dir frag
```
Comment (batrick) [inline src/mds/mdstypes.h:250]: Please undo these whitespace changes. (Use `git reset -p ...`)

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/include/cephfs/sharedtypes.h (hunk 1 above)
Comment (dparmar18) [reply src/include/cephfs/sharedtypes.h:1]: It didn't exist in mdstypes.h, therefore I'm not sure what note to add. Could you please share the LICENSE/copyright notice?

File: src/include/cephfs/sharedtypes.h (hunk 2 above)
Comment (dparmar18) [reply src/include/cephfs/sharedtypes.h:3]: there is already a header named types.h somewhere, won't that be a little mess to have two headers with the same names?

File: src/mds/mdstypes.h (hunk 3 above)
Comment (dparmar18) [reply src/mds/mdstypes.h:250]: I'll do it.

File: src/include/cephfs/sharedtypes.h [hunk 4]
```diff
@@ -0,0 +1,946 @@
...
+#include "include/fs_types.h"
+#include "include/ceph_fs.h"
+
+#include "mds/inode_backtrace.h"
+
+#include <boost/spirit/include/qi.hpp>
+#include <boost/pool/pool.hpp>
+#include "include/ceph_assert.h"
+#include <boost/serialization/strong_typedef.hpp>
+#include "common/ceph_json.h"
```
[the remark this answers is not in the document]
Comment (dparmar18) [reply src/include/cephfs/sharedtypes.h:33]: Not double-checked though, I'll do it in sometime

[the remark this answers is not in the document]
Comment (dparmar18) [reply src/include/cephfs/sharedtypes.h:33]: I double-checked both the files and none of the headers are obsolete.

File: src/include/cephfs/sharedtypes.h (hunk 1 above)
Comment (lxbsz) [reply src/include/cephfs/sharedtypes.h:1]: Please refer to `client/RWRef.h`, this was what I did recently.

File: src/include/cephfs/sharedtypes.h (hunk 2 above)
Comment (lxbsz) [reply src/include/cephfs/sharedtypes.h:3]: IMO it should be okay if there are in different paths.

File: src/include/cephfs/sharedtypes.h (hunk 1 above)
Comment (dparmar18) [reply src/include/cephfs/sharedtypes.h:1]: Okay

File: src/include/cephfs/sharedtypes.h (hunk 2 above)
Comment (gregsfortytwo) [reply src/include/cephfs/sharedtypes.h:13]: > the ifndef will also be CEPH_TYPES_H, I hope it won't be an issue right?

That definitely will be an issue! Would have to stick the path in there or something.

File: src/include/cephfs/sharedtypes.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (dparmar18) [reply src/include/cephfs/sharedtypes.h:43]: Ah, found `mdstypes.h` doesn't use `boost/serialization/strong_typedef.hpp` anymore. There was just one instance where `BOOST_STRONG_TYPEDEF` was used which has been moved to `sharedtypes.h` now

File: src/include/cephfs/sharedtypes.h (hunk 2 above)
Comment (dparmar18) [reply src/include/cephfs/sharedtypes.h:13]: @batrick Can we keep it as sharedypes.h, it also would not lead to ambiguity with the other `types.h` and the name means what it is, i.e. containing sharedtypes?

Comment (dparmar18) [reply src/include/cephfs/sharedtypes.h:13]: That makes sense, I'll rename it.

Comment (dparmar18) [reply src/include/cephfs/sharedtypes.h]: @batrick I've pushed the changes, PTAL

Comment (dparmar18) [reply src/include/cephfs/sharedtypes.h]: BTW the commits are just to have clarity of what changes I have made so far, I'll squash them up after all the requested changes are incorporated.

**Comment by rishabh-d-dave** [comment]
@dparmar18 `Signed-off-by` line is missing in latest commits.

**Comment by dparmar18** [comment]
> @dparmar18 `Signed-off-by` line is missing in latest commits.

Thanks, it's fixed.

File: src/mds/mdstypes.h (hunk 3 above)
Comment (batrick) [reply src/mds/mdstypes.h:249]: I still see these whitespace changes?

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/mds/mdstypes.h (hunk 3 above)
Comment (dparmar18) [reply src/mds/mdstypes.h:249]: It was fixed in the last commit

Comment (dparmar18) [reply src/mds/mdstypes.h]: Done

**Comment by rishabh-d-dave** [comment]
@dparmar18 BTW, rgw label is set on this PR. Is it by mistake?

**Review by batrick — APPROVED** [review]

**Review by neesingh-rh — COMMENTED** [review]
minor nit: newline missing at the end of the `/cephfs/types.h`.Otherwise, LGTM

**Comment by dparmar18** [comment]
> minor nit: newline missing at the end of the `/cephfs/types.h`.Otherwise, LGTM

Yeah added now

**Review by nmshelke — COMMENTED** [review]
LGTM

**Review by vshankar — APPROVED** [review]

**Review by lxbsz — APPROVED** [review]

**Review by nmshelke — APPROVED** [review]

**Comment by rishabh-d-dave** [comment]
Looks good otherwise.

**Comment by lxbsz** [comment]
> > @dparmar18 - https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/63017//consoleFull
> > ```
> > error: Installed (but unpackaged) file(s) found:
> >    /usr/include/cephfs/types.h
> > 
> > 
> > RPM build errors:
> >     Installed (but unpackaged) file(s) found:
> >    /usr/include/cephfs/types.h
> > + rm -fr /tmp/install-deps.999737
> > Build step 'Execute shell' marked build as failure```
> > ```
> 
> @vshankar What spec file do I need to add this file path to? Also what way should I test it after adding the file-path in the respective spec file?

I think you can refer to what the `%{_includedir}/cephfs/metrics/Types.h` does in `ceph.spec.in`.

**Comment by rishabh-d-dave** [comment]
> > > @dparmar18 - https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/63017//consoleFull
> > > ```
> > > error: Installed (but unpackaged) file(s) found:
> > >    /usr/include/cephfs/types.h
> > > 
> > > 
> > > RPM build errors:
> > >     Installed (but unpackaged) file(s) found:
> > >    /usr/include/cephfs/types.h
> > > + rm -fr /tmp/install-deps.999737
> > > Build step 'Execute shell' marked build as failure```
> > > ```
> > 
> > 
> > @vshankar What spec file do I need to add this file path to? Also what way should I test it after adding the file-path in the respective spec file?
> 
> I think you can refer to what the `%{_includedir}/cephfs/metrics/Types.h` does in `ceph.spec.in`.

@lxbsz True. I and Dhairya figured the same yesterday but any idea how to re-run that job to test that change? And is possible to do it locally without too much hassle?

**Comment by lxbsz** [comment]
> > > > @dparmar18 - https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/63017//consoleFull
> > > > ```
> > > > error: Installed (but unpackaged) file(s) found:
> > > >    /usr/include/cephfs/types.h
> > > > 
> > > > 
> > > > RPM build errors:
> > > >     Installed (but unpackaged) file(s) found:
> > > >    /usr/include/cephfs/types.h
> > > > + rm -fr /tmp/install-deps.999737
> > > > Build step 'Execute shell' marked build as failure```
> > > > ```
> > > 
> > > 
> > > @vshankar What spec file do I need to add this file path to? Also what way should I test it after adding the file-path in the respective spec file?
> > 
> > 
> > I think you can refer to what the `%{_includedir}/cephfs/metrics/Types.h` does in `ceph.spec.in`.
> 
> @lxbsz True. I and Dhairya figured the same yesterday but any idea how to re-run that job to test that change? And is possible to do it locally without too much hassle?

Maybe you can test this locally by using the docker, please see `src/test/centos-8/`. Or you can push this PR to the `https://github.com/ceph/ceph-ci`, it should fire a job to build the rpm.

**Comment by dparmar18** [comment]
> > > @dparmar18 - https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/63017//consoleFull
> > > ```
> > > error: Installed (but unpackaged) file(s) found:
> > >    /usr/include/cephfs/types.h
> > > 
> > > 
> > > RPM build errors:
> > >     Installed (but unpackaged) file(s) found:
> > >    /usr/include/cephfs/types.h
> > > + rm -fr /tmp/install-deps.999737
> > > Build step 'Execute shell' marked build as failure```
> > > ```
> > 
> > 
> > @vshankar What spec file do I need to add this file path to? Also what way should I test it after adding the file-path in the respective spec file?
> 
> I think you can refer to what the `%{_includedir}/cephfs/metrics/Types.h` does in `ceph.spec.in`.

@lxbsz I can also see ceph.spec.in for centos. fedora and opensuse under src/test, Do i also need to add the file path to those spec files?

**Comment by lxbsz** [comment]
> > > > @dparmar18 - https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/63017//consoleFull
> > > > ```
> > > > error: Installed (but unpackaged) file(s) found:
> > > >    /usr/include/cephfs/types.h
> > > > 
> > > > 
> > > > RPM build errors:
> > > >     Installed (but unpackaged) file(s) found:
> > > >    /usr/include/cephfs/types.h
> > > > + rm -fr /tmp/install-deps.999737
> > > > Build step 'Execute shell' marked build as failure```
> > > > ```
> > > 
> > > 
> > > @vshankar What spec file do I need to add this file path to? Also what way should I test it after adding the file-path in the respective spec file?
> > 
> > 
> > I think you can refer to what the `%{_includedir}/cephfs/metrics/Types.h` does in `ceph.spec.in`.
> 
> @lxbsz I can also see ceph.spec.in for centos. fedora and opensuse under src/test, Do i also need to add the file path to those spec files?

The others are all link files, you can check it. So no need.

**Comment by dparmar18** [comment]
> > > > > @dparmar18 - https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/63017//consoleFull
> > > > > ```
> > > > > error: Installed (but unpackaged) file(s) found:
> > > > >    /usr/include/cephfs/types.h
> > > > > 
> > > > > 
> > > > > RPM build errors:
> > > > >     Installed (but unpackaged) file(s) found:
> > > > >    /usr/include/cephfs/types.h
> > > > > + rm -fr /tmp/install-deps.999737
> > > > > Build step 'Execute shell' marked build as failure```
> > > > > ```
> > > > 
> > > > 
> > > > @vshankar What spec file do I need to add this file path to? Also what way should I test it after adding the file-path in the respective spec file?
> > > 
> > > 
> > > I think you can refer to what the `%{_includedir}/cephfs/metrics/Types.h` does in `ceph.spec.in`.
> > 
> > 
> > @lxbsz I can also see ceph.spec.in for centos. fedora and opensuse under src/test, Do i also need to add the file path to those spec files?
> 
> The others are all link files, you can check it. So no need.

Thanks!

**Comment by dparmar18** [comment]
> > > > > @dparmar18 - https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/63017//consoleFull
> > > > > ```
> > > > > error: Installed (but unpackaged) file(s) found:
> > > > >    /usr/include/cephfs/types.h
> > > > > 
> > > > > 
> > > > > RPM build errors:
> > > > >     Installed (but unpackaged) file(s) found:
> > > > >    /usr/include/cephfs/types.h
> > > > > + rm -fr /tmp/install-deps.999737
> > > > > Build step 'Execute shell' marked build as failure```
> > > > > ```
> > > > 
> > > > 
> > > > @vshankar What spec file do I need to add this file path to? Also what way should I test it after adding the file-path in the respective spec file?
> > > 
> > > 
> > > I think you can refer to what the `%{_includedir}/cephfs/metrics/Types.h` does in `ceph.spec.in`.
> > 
> > 
> > @lxbsz True. I and Dhairya figured the same yesterday but any idea how to re-run that job to test that change? And is possible to do it locally without too much hassle?
> 
> Maybe you can test this locally by using the docker, please see `src/test/centos-8/`. Or you can push this PR to the `https://github.com/ceph/ceph-ci`, it should fire a job to build the rpm.

Okay, but the build process would again take a few hours too, right?

**Comment by lxbsz** [comment]
...
> > 
> > Maybe you can test this locally by using the docker, please see `src/test/centos-8/`. Or you can push this PR to the `https://github.com/ceph/ceph-ci`, it should fire a job to build the rpm.
> 
> Okay, but the build process would again take a few hours too, right?

It seems.

**Comment by dparmar18** [comment]
@vshankar looks like the builds are getting generated successfully in https://shaman.ceph.com/builds/ceph/ after I pushed my latest changes. Build with config that failed previously i.e. centos-x86_64 is built successfully now. PS: build name is 'wip-mds_split'

**Review by vshankar — APPROVED** [review]

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-June-20

**Pull request merged**